gobblin-runtime/src/main/java/org/apache/gobblin/runtime/api/SchedulerLeaseDeterminationStore.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.gobblin.runtime.api;
+
+import java.io.IOException;
+import java.sql.Timestamp;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Interface defines the two basic actions required for lease determination for each FlowActionType event for a flow.
+ * It is used by the {@link SchedulerLeaseAlgoHandler} to allow multiple scheduler's on different hosts to determine
+ * which scheduler is tasked with ensuring the FlowAction is taken for the trigger.
+ */
+public interface SchedulerLeaseDeterminationStore {
+  static final Logger LOG = LoggerFactory.getLogger(SchedulerLeaseDeterminationStore.class);
+
+  // Enum is used to reason about the three possible scenarios that can result from an attempt to obtain a lease for a
+  // particular trigger event of a flow
+  enum LeaseAttemptStatus {
+    LEASE_OBTAINED,
+    PREVIOUS_LEASE_EXPIRED,
+    PREVIOUS_LEASE_VALID
```

Comment:
I respect using an enum for the tri-state result, but the three I had imagined are more like:
* LEASE_IS_YOURS / LEASE_OBTAINED (acquired)
* LEASED_TO_ANOTHER / LEASE_NOT_AVAILABLE
* NO_LONGER_LEASING / LEASE_NOT_POSSIBLE (not necessary / there's nothing to lease)

did you have the same three concepts in mind, just named differently--or different idea here?

Reply:
those are the similar ideas, although I'm not sure what the second and third states exactly entail in the states you described above. For the second state it tells us the lease is given to another but not sure if another host is in the process of pursuing or completed lease. What is the third case? Is that if another host completed pursuing the lease?

I was thinking having the states represent the next course of action. 
```
Case 1: lease is acquired completely by itself or another host (do nothing after)
Case 2: lease attempted by another expired (can try to get lease now)
Case 3: previous lease attempt still ongoing (set a reminder to check back on lease)
```

Reply:
we're thinking similarly, but I see a few diffs:
a. the most consequential discernment is whether the current host acquired the lease (hence must be separated out so THE ONLY possibility for case 1)
b. the "linger" timeout should not be so short that the current host would fail to acquire, but at that same moment discover that whoever else "has" the lease has actually exceeded the leases expiration.  each host must defer within the "linger" period, and should only attempt to acquire after that expires.  if the host should fail because another won the race to acquire it, that newly acquired lease should be nowhere near expiration.
c. with NO_LONGER_LEASING, I sought to capture the termination case, where a host realizes the trigger event has been fully handled, so there's neither a need to try to acquire nor even set a reminder.  in fact, under normal operating conditions, such a self-timer should regularly conclude by the host determining "there's nothing to lease", because the trigger event has been fully handled.

Reply:
I was also debating the cases you're identifying here particularly distinguishing whether the current host succeeded or the lease has gone to another. In my original code I expected the class implementing this interface to actually incorporate all the logic for setting a reminder event or persisting the flow trigger to the DagActionStore so the logic was less general. The abstraction you've described makes more sense as I generalize this to a `LeaseArbiter` for any flow event, so changing the enums to this.